stable_baselines3/her/her_replay_buffer.py

```diff
@@ -185,347 +116,235 @@ def set_env(self, env: VecEnv) -> None:
+        :param env_indices: Indices of the envrionments
+        :param env: associated gym VecEnv to normalize the
+            observations/rewards when sampling, defaults to None
+        :return: Samples
         """
-        # `maybe_vec_env=None` as we should store unnormalized transitions,
-        # they will be normalized at sampling time
-        return self._sample_transitions(
-            batch_size=None,
-            maybe_vec_env=None,
-            online_sampling=False,
-            n_sampled_goal=n_sampled_goal,
-        )  # pytype: disable=bad-return-type
-
-    def sample_goals(
-        self,
-        episode_indices: np.ndarray,
-        her_indices: np.ndarray,
-        transitions_indices: np.ndarray,
-    ) -> np.ndarray:
+        # Normalize if needed and remove extra dimension (we are using only one env for now)
+        obs_ = self._normalize_obs({key: obs[batch_inds, env_indices, :] for key, obs in self.observations.items()}, env)
+        next_obs_ = self._normalize_obs(
+            {key: obs[batch_inds, env_indices, :] for key, obs in self.next_observations.items()}, env
+        )
+
+        # Convert to torch tensor
+        observations = {key: self.to_torch(obs) for key, obs in obs_.items()}
+        next_observations = {key: self.to_torch(obs) for key, obs in next_obs_.items()}
+
+        return DictReplayBufferSamples(
+            observations=observations,
+            actions=self.to_torch(self.actions[batch_inds, env_indices]),
+            next_observations=next_observations,
+            # Only use dones that are not due to timeouts
+            # deactivated by default (timeouts is initialized as an array of False)
+            dones=self.to_torch(self.dones[batch_inds, env_indices] * (1 - self.timeouts[batch_inds, env_indices])).reshape(
+                -1, 1
+            ),
+            rewards=self.to_torch(self._normalize_reward(self.rewards[batch_inds, env_indices].reshape(-1, 1), env)),
+        )
+
+    def _get_virtual_samples(
+        self, batch_inds: np.ndarray, env_indices: np.ndarray, env: Optional[VecNormalize] = None
+    ) -> DictReplayBufferSamples:
+        """
+        Get the samples, sample new desired goals and compute new rewards.
+
+        :param batch_inds: Indices of the transitions
+        :param env_indices: Indices of the envrionments
+        :param env: associated gym VecEnv to normalize the
+            observations/rewards when sampling, defaults to None
+        :return: Samples, with new desired goals and new rewards
+        """
+        # Get infos and obs
+        obs = {key: obs[batch_inds, env_indices, :] for key, obs in self.observations.items()}
+        next_obs = {key: obs[batch_inds, env_indices, :] for key, obs in self.next_observations.items()}
+        # For speed up we replace this
+        # infos = copy.deepcopy(self.infos[batch_inds, env_indices])
+        # by
+        infos = [{} for _ in range(len(batch_inds))]
```

Comment:
This is an omission. I noticed that the info dict deepcopy was taking a non-negligible amount of time so I removed it because the info is rarely (never to my knowledge) used to compute the reward. I'll put it back and see what it implies in terms of time.

Reply:
then, maybe add an option to activate it?
(I used it in the past to compute continuity cost for instance)

Reply:
```python
import panda_gym

from stable_baselines3 import SAC, HerReplayBuffer
from stable_baselines3.common.env_util import make_vec_env

env = make_vec_env("PandaPush-v1", n_envs=4)

SAC("MultiInputPolicy", env, replay_buffer_class=HerReplayBuffer, verbose=1, learning_starts=1_000).learn(20_000)
```

| info deepcopy | with | without |
|-|-|-|
| FPS | 249 | 257 |

Reply:
I added an argument for it, disabled by default